src/v/cloud_storage/topic_manifest_downloader.cc

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "cloud_storage/topic_manifest_downloader.h"
+
+#include "base/outcome.h"
+#include "cloud_storage/topic_manifest.h"
+#include "cloud_storage/topic_path_utils.h"
+#include "cloud_storage/types.h"
+#include "cloud_storage_clients/client.h"
+#include "container/fragmented_vector.h"
+#include "hashing/xx.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/lowres_clock.hh>
+
+namespace cloud_storage {
+
+namespace {
+bool bin_manifest_filter(
+  const cloud_storage_clients::client::list_bucket_item& i) {
+    return i.key.ends_with("topic_manifest.bin");
+};
+} // namespace
+topic_manifest_downloader::topic_manifest_downloader(
+  const cloud_storage_clients::bucket_name bucket,
+  std::optional<ss::sstring> hint,
+  const model::topic_namespace topic,
+  remote& remote)
+  : bucket_(bucket)
+  , label_hint_(std::move(hint))
+  , topic_(topic)
+  , remote_(remote) {}
+
+ss::future<result<find_topic_manifest_outcome, error_outcome>>
+topic_manifest_downloader::download_manifest(
+  retry_chain_node& parent_retry,
+  ss::lowres_clock::time_point deadline,
+  model::timestamp_clock::duration backoff,
+  topic_manifest* manifest) {
+    retry_chain_node retry_node(deadline, backoff, &parent_retry);
+
+    // First look for topic manifests with a label.
+    ss::sstring remote_label_str = label_hint_.value_or("");
+    const auto labeled_manifest_filter = fmt::format(
+      "{}/{}", labeled_topic_manifest_root(topic_), remote_label_str);
+    auto list_res = co_await remote_.list_objects(
+      bucket_,
+      retry_node,
+      cloud_storage_clients::object_key{labeled_manifest_filter},
+      std::nullopt,
+      bin_manifest_filter);
```

Comment:
does this filter have a cluster id injected in it?

Reply:
We chatted in person, bringing it back here:

Yeah, the expectation is that a caller may pass in a hint that would potentially include the cluster UUID